include/caffe/solver.hpp

```diff
@@ -129,6 +129,28 @@ class AdaGradSolver : public SGDSolver<Dtype> {
 };
 
 template <typename Dtype>
+class AdamSolver : public SGDSolver<Dtype> {
+ public:
+  explicit AdamSolver(const SolverParameter& param)
+      : SGDSolver<Dtype>(param) { constructor_sanity_check(); initAdam();}
+  explicit AdamSolver(const string& param_file)
+      : SGDSolver<Dtype>(param_file) { constructor_sanity_check(); initAdam();}
+
+ protected:
+  void initAdam();
+  virtual void ComputeUpdateValue(int param_id, Dtype rate);
+  void constructor_sanity_check() {
+    CHECK_EQ(0, this->param_.momentum())
+        << "Momentum cannot be used with Adam.";
+    CHECK_NE(this->param_.stepsize(), 0) << "Adam needs a non-zero stepsize.";
+  }
+
+  vector<shared_ptr<Blob<Dtype> > > val_v_, val_m_, val_t_;
```

Comment:
You can reuse the history_ and temp_ if the SGD solver and just add a val_m_ (or history2_) here.